LocalVariablesSorter only accepts expanded frames error when using Quarkus Log in Panache entities

### Describe the bug

Having a Panache Entity defined as the following: 

```java
package org.acme;

import javax.persistence.Entity;

import io.quarkus.hibernate.orm.panache.PanacheEntity;
import io.quarkus.logging.Log;

@Entity
public class MyEntity extends PanacheEntity {
    
    public void something() {
        Log.info("Hi!");
    }
}
```

The following error happens when building the project:

```java
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:2.5.2.Final:build (default) on project fancyproject: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]     [error]: Build step io.quarkus.deployment.steps.ClassTransformingBuildStep#handleClassTransformation threw an exception: java.lang.IllegalStateException: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: LocalVariablesSorter only accepts expanded frames (see ClassReader.EXPAND_FRAMES)
[ERROR]     at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:896)
[ERROR]     at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR]     at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]     at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
[ERROR]     at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
[ERROR]     at java.base/java.lang.Thread.run(Thread.java:833)
[ERROR]     at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: LocalVariablesSorter only accepts expanded frames (see ClassReader.EXPAND_FRAMES)
[ERROR]     at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
[ERROR]     at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
[ERROR]     at io.quarkus.deployment.steps.ClassTransformingBuildStep.handleClassTransformation(ClassTransformingBuildStep.java:231)
[ERROR]     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR]     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
[ERROR]     at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR]     at java.base/java.lang.reflect.Method.invoke(Method.java:568)
[ERROR]     at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:887)
[ERROR]     ... 6 more
[ERROR] Caused by: java.lang.IllegalArgumentException: LocalVariablesSorter only accepts expanded frames (see ClassReader.EXPAND_FRAMES)
[ERROR]     at org.objectweb.asm.commons.LocalVariablesSorter.visitFrame(LocalVariablesSorter.java:189)
[ERROR]     at org.objectweb.asm.ClassReader.readCode(ClassReader.java:2075)
[ERROR]     at org.objectweb.asm.ClassReader.readMethod(ClassReader.java:1514)
[ERROR]     at org.objectweb.asm.ClassReader.accept(ClassReader.java:744)
[ERROR]     at org.objectweb.asm.ClassReader.accept(ClassReader.java:424)
[ERROR]     at io.quarkus.hibernate.orm.deployment.HibernateEntityEnhancer$HibernateEnhancingClassVisitor.visitEnd(HibernateEntityEnhancer.java:81)
[ERROR]     at org.objectweb.asm.ClassVisitor.visitEnd(ClassVisitor.java:377)
[ERROR]     at io.quarkus.panache.common.deployment.visitors.PanacheEntityClassAccessorGenerationVisitor.visitEnd(PanacheEntityClassAccessorGenerationVisitor.java:111)
[ERROR]     at org.objectweb.asm.ClassVisitor.visitEnd(ClassVisitor.java:377)
[ERROR]     at org.objectweb.asm.ClassVisitor.visitEnd(ClassVisitor.java:377)
[ERROR]     at io.quarkus.panache.common.deployment.visitors.PanacheEntityClassOperationGenerationVisitor.visitEnd(PanacheEntityClassOperationGenerationVisitor.java:121)
[ERROR]     at org.objectweb.asm.ClassReader.accept(ClassReader.java:748)
[ERROR]     at org.objectweb.asm.ClassReader.accept(ClassReader.java:424)
[ERROR]     at io.quarkus.deployment.steps.ClassTransformingBuildStep.transformClass(ClassTransformingBuildStep.java:317)
[ERROR]     at io.quarkus.deployment.steps.ClassTransformingBuildStep$2.call(ClassTransformingBuildStep.java:204)
[ERROR]     at io.quarkus.deployment.steps.ClassTransformingBuildStep$2.call(ClassTransformingBuildStep.java:191)
[ERROR]     at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[ERROR]     at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[ERROR]     at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[ERROR]     at java.base/java.lang.Thread.run(Thread.java:833)
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

### Expected behavior

No build errors

### Actual behavior

Build error

### How to Reproduce?

1. Create a project using hibernate-orm-panache (and any jdbc extension): https://code.quarkus.io/d?e=hibernate-orm-panache&e=jdbc-postgresql&cn=code.quarkus.io
2. Create an entity as in this issue description
3. Run `mvn clean package`

### Output of `uname -a` or `ver`

Fedora 35

### Output of `java -version`

JDK 17

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.5.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.8.4

### Additional information

_No response_

--- a/core/deployment/src/main/java/io/quarkus/deployment/QuarkusClassVisitor.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/QuarkusClassVisitor.java
@@ -1,0 +1,32 @@
+package io.quarkus.deployment;
+
+import org.objectweb.asm.ClassVisitor;
+
+/**
+ * A subclass of {@link ClassVisitor} that allows carrying around data
+ * that are useful in the context of Quarkus bytecode transformations.
+ * Class visitors that require access to these data should extend this class.
+ */
+public abstract class QuarkusClassVisitor extends ClassVisitor {
+    public QuarkusClassVisitor(int api) {
+        super(api);
+    }
+
+    public QuarkusClassVisitor(int api, ClassVisitor classVisitor) {
+        super(api, classVisitor);
+    }
+
+    // ---
+
+    // this could possibly be a Map<String, Object> or something like that,
+    // but there's no need at the moment
+    private int originalClassReaderOptions;
+
+    public int getOriginalClassReaderOptions() {
+        return originalClassReaderOptions;
+    }
+
+    public void setOriginalClassReaderOptions(int originalClassReaderOptions) {
+        this.originalClassReaderOptions = originalClassReaderOptions;
+    }
+}

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/ClassTransformingBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/ClassTransformingBuildStep.java
@@ -32,6 +32,7 @@ import org.objectweb.asm.ClassWriter;
 import io.quarkus.bootstrap.BootstrapDebug;
 import io.quarkus.bootstrap.classloading.ClassPathElement;
 import io.quarkus.bootstrap.classloading.QuarkusClassLoader;
+import io.quarkus.deployment.QuarkusClassVisitor;
 import io.quarkus.deployment.QuarkusClassWriter;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.ApplicationArchivesBuildItem;
@@ -313,6 +314,9 @@ public class ClassTransformingBuildStep {
             ClassVisitor visitor = writer;
             for (BiFunction<String, ClassVisitor, ClassVisitor> i : visitors) {
                 visitor = i.apply(className, visitor);
+                if (visitor instanceof QuarkusClassVisitor) {
+                    ((QuarkusClassVisitor) visitor).setOriginalClassReaderOptions(classReaderOptions);
+                }
             }
             cr.accept(visitor, classReaderOptions);
             data = writer.toByteArray();

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java
@@ -10,6 +10,7 @@ import org.objectweb.asm.ClassReader;
 import org.objectweb.asm.ClassVisitor;
 import org.objectweb.asm.ClassWriter;
 
+import io.quarkus.deployment.QuarkusClassVisitor;
 import io.quarkus.deployment.QuarkusClassWriter;
 import io.quarkus.gizmo.Gizmo;
 import net.bytebuddy.ClassFileVersion;
@@ -36,7 +37,7 @@ public final class HibernateEntityEnhancer implements BiFunction<String, ClassVi
         return new HibernateEnhancingClassVisitor(className, outputClassVisitor);
     }
 
-    private static class HibernateEnhancingClassVisitor extends ClassVisitor {
+    private static class HibernateEnhancingClassVisitor extends QuarkusClassVisitor {
 
         private final String className;
         private final ClassVisitor outputClassVisitor;
@@ -78,7 +79,7 @@ public final class HibernateEntityEnhancer implements BiFunction<String, ClassVi
             final byte[] transformedBytes = hibernateEnhancement(className, inputBytes);
             //Then re-convert the transformed bytecode to not interrupt the visitor chain:
             ClassReader cr = new ClassReader(transformedBytes);
-            cr.accept(outputClassVisitor, 0);
+            cr.accept(outputClassVisitor, super.getOriginalClassReaderOptions());
         }
 
         private byte[] hibernateEnhancement(final String className, final byte[] originalBytes) {

--- a/integration-tests/logging-panache/pom.xml
+++ b/integration-tests/logging-panache/pom.xml
@@ -20,14 +20,14 @@
             <artifactId>quarkus-arc-deployment</artifactId>
             <scope>test</scope>
         </dependency>
-        <!--
-          - this dependency is enough to test that the Logging with Panache
-          - and Hibernate ORM with Panache class transformations do not collide
-          - on setting the class reader options (issue #20569)
-          -->
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-hibernate-orm-panache</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-jdbc-h2</artifactId>
             <scope>test</scope>
         </dependency>
         <dependency>
@@ -56,6 +56,19 @@
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-hibernate-orm-panache-deployment</artifactId>
+            <version>${project.version}</version>
+            <type>pom</type>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-jdbc-h2-deployment</artifactId>
             <version>${project.version}</version>
             <type>pom</type>
             <scope>test</scope>

--- a/integration-tests/logging-panache/src/test/java/io/quarkus/logging/LoggingEntity.java
+++ b/integration-tests/logging-panache/src/test/java/io/quarkus/logging/LoggingEntity.java
@@ -1,0 +1,12 @@
+package io.quarkus.logging;
+
+import javax.persistence.Entity;
+
+import io.quarkus.hibernate.orm.panache.PanacheEntity;
+
+@Entity
+public class LoggingEntity extends PanacheEntity {
+    public void something() {
+        Log.info("Hi!");
+    }
+}

--- a/integration-tests/logging-panache/src/test/java/io/quarkus/logging/LoggingWithPanacheTest.java
+++ b/integration-tests/logging-panache/src/test/java/io/quarkus/logging/LoggingWithPanacheTest.java
@@ -18,10 +18,10 @@ public class LoggingWithPanacheTest {
     @RegisterExtension
     static final QuarkusUnitTest test = new QuarkusUnitTest()
             .withApplicationRoot((jar) -> jar
-                    .addClasses(LoggingBean.class, NoStackTraceTestException.class))
+                    .addClasses(LoggingBean.class, LoggingEntity.class, NoStackTraceTestException.class))
             .overrideConfigKey("quarkus.log.category.\"io.quarkus.logging\".min-level", "TRACE")
             .overrideConfigKey("quarkus.log.category.\"io.quarkus.logging\".level", "TRACE")
-            .setLogRecordPredicate(record -> "io.quarkus.logging.LoggingBean".equals(record.getLoggerName()))
+            .setLogRecordPredicate(record -> record.getLoggerName().startsWith("io.quarkus.logging.Logging"))
             .assertLogRecords(records -> {
                 Formatter formatter = new PatternFormatter("[%p] %m");
                 List<String> lines = records.stream().map(formatter::format).map(String::trim).collect(Collectors.toList());
@@ -38,7 +38,8 @@ public class LoggingWithPanacheTest {
                         "[WARN] three: foo | bar | baz",
                         "[ERROR] four: foo | bar | baz | quux",
                         "[WARN] foo | bar | baz | quux: io.quarkus.logging.NoStackTraceTestException",
-                        "[ERROR] Hello Error: io.quarkus.logging.NoStackTraceTestException");
+                        "[ERROR] Hello Error: io.quarkus.logging.NoStackTraceTestException",
+                        "[INFO] Hi!");
             });
 
     @Inject
@@ -47,5 +48,6 @@ public class LoggingWithPanacheTest {
     @Test
     public void test() {
         bean.doSomething();
+        new LoggingEntity().something();
     }
 }

